feat: Improve tab focus on clear input with global config #2410

adambennett opened the pull request:
This PR further improves on the feature added in #2546.

In #2546, we added a property to allow users to configure the clear icon tab behavior on a per-control basis. This PR expands on the concept by allowing developers to disable the behavior at the application level by passing a flag to NgSelectConfig.

Under the new implementation, the behavior works as follows:
- Config flag is true, user does not provide [tabFocusOnClearButton] property on the control = Focus (DEFAULT)
- Config flag is true, user provides [tabFocusOnClearButton] as 'false' = NO focus
- Config flag is true, user provides [tabFocusOnClearButton] as 'true' = Focus
- Config flag is false, user does not provide [tabFocusOnClearButton] property on the control = NO focus
- Config flag is false, user provides [tabFocusOnClearButton] as 'false' = NO focus
- Config flag is false, user provides [tabFocusOnClearButton] as 'true' = Focus

To cover each of these cases I have added 6 new test cases. Similar to #2545 I modified the organization of the test cases slightly to allow for modification of NgSelectConfig for these tests. A new test section called "keyboard events (tab)" has been added and all tab tests from the original "keyboard events" section have been moved here.

This PR further expands on issue #2410. @pavankjadda please let me know if you have any feedback, thank you!

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -117,7 +117,7 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 	@Input() inputAttrs: { [key: string]: string } = {};
 	@Input({ transform: numberAttribute }) tabIndex: number;
-	tabFocusOnClearButton = input(true, { transform: booleanAttribute });
+	@Input({ transform: booleanAttribute }) tabFocusOnClearButton?: boolean;
```
Make it signal

adambennett replied:
This has been updated

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -168,6 +168,7 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 	element: HTMLElement;
 	focused: boolean;
 	escapeHTML = true;
+	finalTabFocus = true;
```
Delete this and use `_mergeGlobalConfig` method to merge both

adambennett replied:
I think I need another variable here, and also have to call the update on ngOnChanges since the property might be different for each dropdown. If I make it one value that gets set during the mergeGlobalConfig, it can never get reset when you move to the next dropdown

pavankjadda replied:
Follow [similar pattern](https://github.com/ng-select/ng-select/blob/master/src/ng-select/lib/ng-select.component.ts#L1048). Define private variable like `_tabFocusOnClear` and store derived value in that.

adambennett replied:
Is this ok now? I changed it to a private variable but I did also need to add a getter so that I can fetch the value in the html. Let me know if you think there's a better approach, thanks!

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -319,6 +320,7 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 	ngOnInit() {
 		this._handleKeyPresses();
 		this._setInputAttributes();
+		this._setFinalTabFocus();
```
Same as above

adambennett replied:
I moved this to the mergeGlobalConfig function

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -874,6 +877,12 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		}
 	}
 
+	private _setFinalTabFocus() {
```
Same as above

adambennett replied:
Re-using in mergeGlobalConfig and in ngOnChanges

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -339,6 +341,7 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		if (changes.inputAttrs) {
 			this._setInputAttributes();
 		}
+		this._setFinalTabFocus();
```
Same as above

adambennett replied:
Needed, see comment above

adambennett commented:
Hi @pavankjadda, are you still looking for some changes on this PR? I believe I addressed all your comments above as best as possible without breaking the functionality I am trying to achieve, but please let me know if there's still anything I can improve. Thanks!
